### ISSUE TITLE
Running superlinter on self hosted github runner is failing

### ISSUE THREAD

**Issue by moble-joseph** [github/super-linter#1536]
We were planning to use a self-hosted GitHub runner for a few of our projects. 
So took a Ubuntu machine and installed the github runner. Everything went fine except superlinter.
Superlinter failed with the below error message. Tried setting environment variables and all, but didn't worked. What is going wrong here?

2021-05-10 09:31:14 [INFO]   Gathering user validation information...
2021-05-10 09:31:14 [INFO]   - Only validating [new], or [edited] files in code base...
2021-05-10 09:31:20 [INFO]   Moved cert into location, adding to trust store...
2021-05-10 09:31:20 [INFO]   Successfully added cert to trust store
fatal: not a git repository (or any parent up to mount point /github)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
2021-05-10 09:31:20 [INFO]   Failed to switch to master branch to get files changed!
2021-05-10 09:31:20 [FATAL]   [fatal: not a git repository (or any parent up to mount point /github)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).]

**Comment by admiralAwkbar** [comment]
hey @moble-joseph 

Can you try to run with the flag:
```bash
- name: Lint Code Base
  uses: github/super-linter@v3
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    ACTIONS_RUNNER_DEBUG: true
```

And we can see whats going on?

it seems like its in the `BuildFileList` function, and having issues seeing the code..

**Comment by moble-joseph** [comment]
@admiralAwkbar Ok sure, let me try and see what it says

**Comment by moble-joseph** [comment]
@admiralAwkbar This error happens only when we give the flag
VALIDATE_ALL_CODEBASE: false
When I remove that, it works fine.

**Comment by zkoppert** [comment]
I noticed `Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).` in your error log. Are you putting the repo on a separate filesystem from where the action is running?

**Comment by moble-joseph** [comment]
@zkoppert No it's on the same VM and same working folder. Even i tried setting an environment variable
GIT_DISCOVERY_ACROSS_FILESYSTEM: true
But it didn't work out.

**Comment by ferrarimarco** [comment]
```
fatal: not a git repository (or any parent up to mount point /github)
```

Likely the issue is not `GIT_DISCOVERY_ACROSS_FILESYSTEM`, but the fact that there's no .git directory there.

Can you enable debug output (as @admiralAwkbar suggested above) and post the value of the `VALIDATE_ALL_CODEBASE` variable? I'd also try to explicitly set that to `true`?

@admiralAwkbar: I see that you commented vars in `linter.sh` and other places (https://github.com/github/super-linter/pull/1516). Do you remember why?

**Comment by moble-joseph** [comment]
@ferrarimarco @zkoppert 
Attached is the log file. 
[Runner Log.txt](https://github.com/github/super-linter/files/6478990/Runner.Log.txt)

**Comment by admiralAwkbar** [comment]
@ferrarimarco it was a new rule on shellcheck that basically hates self declare vars like:
```bash
SOME_VAR="${SOME_VAR}"
```

If they are the exact same, there was no reason other than to show that var... I'm not a huge fan of that rule...

**Comment by ferrarimarco** [comment]
@moble-joseph from your log:

```
2021-05-14T12:54:50.6526628Z ##[group]Getting Git version info
2021-05-14T12:54:50.6528438Z Working directory is '/root/actions-runner/_work/alt-api/alt-api'
2021-05-14T12:54:50.6529312Z [command]/usr/bin/git version
2021-05-14T12:54:50.6662289Z git version 2.17.1
2021-05-14T12:54:50.6701320Z ##[endgroup]
2021-05-14T12:54:50.6710463Z Deleting the contents of '/root/actions-runner/_work/alt-api/alt-api'
2021-05-14T12:54:50.9101896Z The repository will be downloaded using the GitHub REST API
2021-05-14T12:54:50.9103471Z To create a local Git repository instead, add Git 2.18 or higher to the PATH
2021-05-14T12:54:50.9154800Z Downloading the archive
2021-05-14T12:54:51.8017918Z Writing archive to disk
2021-05-14T12:54:51.8063126Z Extracting the archive
2021-05-14T12:54:51.8108728Z [command]/bin/tar xz -C /root/actions-runner/_work/alt-api/alt-api/d9270dd8-be84-4e15-8454-42567d5479d2 -f /root/actions-runner/_work/alt-api/alt-api/d9270dd8-be84-4e15-8454-42567d5479d2.tar.gz
2021-05-14T12:54:51.8360939Z Resolved version logi-alt-api-9da9fef681ca1a1de8e6a2a587b8e457ed5d5b27
```

TLDR: The runner is downloading an archive instead of cloning the repo, hence it's not a Git repository. Try doing as the message suggests (update Git to 2.18) and try again.

@admiralAwkbar maybe we should add safeguards to fail early when the working directory is not a Git directory?

**Comment by TemekRoman** [comment]
I'm having the same issue with running SuperLinter `v3.17.1` locally:
```
docker run \
	  --rm \
	  -e RUN_LOCAL=true \
	  -e VALIDATE_MARKDOWN=true \
	  -v /home/user/git-repos/wsp/Services/Audit:/tmp/lint \
	  github/super-linter:v3.17.1
```
Here, `/home/user/git-repos/wsp` is the root of the repo. I'm mapping just a part of the repo to the container.
This fails with:
```
fatal: not a git repository (or any parent up to mount point /tmp)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

I found that the issue was introduced in `v3.15.0`.
The provided scenario works fine with SuperLinter `v3.14.5`.

**Comment by jbouganim-glenair** [comment]
running into the same issue. using v3.14.5 didnt fix the issue for me.  Setting `VALIDATE_ALL_CODEBASE` to true still generates the error but for some reason it doesn't cause the step to fail

debug log attached
[5_Lint Code Base.txt](https://github.com/github/super-linter/files/6535729/5_Lint.Code.Base.txt)

yaml:

```
name: Branch Build

on:
  push:
    branches-ignore: [main, dev]
  pull_request:
    branches-ignore: [main, dev]
  workflow_dispatch:

jobs:
  build:
    runs-on: [self-hosted, docker]
    container: python:3.9-alpine
    env:
      CI: false #necessary to NOT treat warnings as errors
    steps:
      - uses: actions/checkout@v2
      - name: Lint Code Base
        uses: github/super-linter@v3
        env:
          VALIDATE_ALL_CODEBASE: false
          DEFAULT_BRANCH: dev
          ACTIONS_RUNNER_DEBUG: true
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      - uses: actions/upload-artifact@v2
        with:
          name: Package
          path: ./
```

**Comment by nhojt** [comment]
I'm also hitting this issue when mapping `/tmp/lint` to part of the repo with no `.git` folder (which worked fine in 3.14.5):

```
---------------------------------------------
--- GitHub Actions Multi Language Linter ----
 - Image Creation Date:[2021-06-03T14:49:09Z]
 - Image Revision:[30812b0716e4b4db1ef465c48cca0f56fa6d33f3]
 - Image Version:[30812b0716e4b4db1ef465c48cca0f56fa6d33f3]
---------------------------------------------
---------------------------------------------
The Super-Linter source code can be found at:
 - https://github.com/github/super-linter
---------------------------------------------
--------------------------------------------
Gathering GitHub information...
NOTE: ENV VAR [RUN_LOCAL] has been set to:[true]
bypassing GitHub Actions variables...
Linting all files in mapped directory:[/tmp/lint]
Successfully found:[GITHUB_TOKEN]
--------------------------------------------
Gathering user validation information...
fatal: not a git repository (or any parent up to mount point /tmp)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
No files were found in the GITHUB_WORKSPACE:[/tmp/lint] to lint!
fatal: not a git repository (or any parent up to mount point /tmp)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
---------------------------------
------ File list to check: ------
---------------------------------
----------------------------------------------
Successfully gathered list of files...
----------------------------------------------
----------------------------------------------
The script has completed
----------------------------------------------
----------------------------------------------
  All file(s) linted successfully with no errors detected
----------------------------------------------
```

**Comment by admiralAwkbar** [comment]
I know we implemented a check to make sure we were in a git repository by looking for a `.git` folder in the root. That seems to be some of the issues some of you have come across. That was meant to make sure we were only looking and validating files that were actively tracked in `git`. As for the mount space issues, that's an interesting one that I'm not quite sure where it stems from...

Im open for options such as a flag to scan everything we find?

**Comment by nhojt** [comment]
Hi @admiralAwkbar that sounds good.

**Comment by nhojt** [comment]
Hi @admiralAwkbar, tested on 4.2.2 with ACTIONS_RUNNER_DEBUG=true and the code is successfully finding the files using the find algo, but then it hits line 819 in buildFileList.sh, tries to execute git status to get the git ignored files list and fails:

```
[DEBUG]   Validate all code base: true...
[DEBUG]   TEST_CASE_RUN: false...
[DEBUG]   ANSIBLE_DIRECTORY: /tmp/lint/ansible...
[DEBUG]   IGNORE_GITIGNORED_FILES: false...
[DEBUG]   USE_FIND_ALGORITHM: true...
[DEBUG]   ----------------------------------------------
[DEBUG]   Populating the file list with all the files in the /tmp/lint workspace using FIND algorithm
[DEBUG]   Loading the files list that Git ignores...
fatal: not a git repository (or any parent up to mount point /tmp)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
[DEBUG]   GIT_IGNORED_FILES contents: 
[DEBUG]   --- GIT_IGNORED_FILES_INDEX contents ---
```

**Comment by jfritz** [comment]
> That was meant to make sure we were only looking and validating files that were actively tracked in `git`. 
> 
> Im open for options such as a flag to scan everything we find?

Note for future people looking for information on running super-linter locally in concert with the above bug: to realize the above change (and therefore avoid an issue where locally linting a specific file (not folder) mounted via a `-v` mount point that does not have `.git` in it) -- you'll have to add `-e USE_FIND_ALGORITHM=true` to bypass the `.git` check so the superlinter can lint your files. See https://github.com/github/super-linter/pull/1640

**Comment by admiralAwkbar** [comment]
@jfritz thanks for the clarification on that!

**Comment by texonidas** [comment]
For anyone else who landed here and wasn't helped by the above:

Your runner needs to have Git installed. If it doesn't, the checkout action will fetch an archive using the GitHub API instead of cloning the repo, which means the container does not receive a repo.
